libvmi/driver/kvm/kvm.c

```diff
@@ -859,22 +858,60 @@ status_t
 kvm_pause_vm(
     vmi_instance_t vmi)
 {
+    unsigned int expect_pause_count = 0;
+
     kvm_instance_t *kvm = kvm_get_instance(vmi);
 
-    if (-1 == kvm->libvirt.virDomainSuspend(kvm->dom)) {
+    // pause vcpus
+    if (kvmi_pause_all_vcpus(kvm->kvmi_dom, &expect_pause_count)) {
+        dbprint(VMI_DEBUG_KVM, "--Failed to pause domain\n");
         return VMI_FAILURE;
     }
+
+    // wait to receive pause events
+    unsigned int pause_events_count = 0;
+    while (pause_events_count < expect_pause_count) {
+        struct kvmi_dom_event *ev = NULL;
+        unsigned int ev_id = 0;
+
+        // wait
+        if (kvmi_wait_event(kvm->kvmi_dom, 30 * 1024)) {
+            dbprint(VMI_DEBUG, "--Failed to receive event\n");
+            return VMI_FAILURE;
+        }
+        // pop
+        if (kvmi_pop_event(kvm->kvmi_dom, &ev)) {
+            dbprint(VMI_DEBUG, "--Failed to pop event\n");
+            return VMI_FAILURE;
+        }
+        // handle event
+        ev_id = ev->event.common.event;
+        switch (ev_id) {
+            case KVMI_EVENT_PAUSE_VCPU:
+                dbprint(VMI_DEBUG_KVM, "--Received VCPU pause event\n");
+                pause_events_count++;
+                free(ev);
```

Comment:
Do I have the responsability to `free` the event structure ?
As `kvmi_pop_event` takes a `event **`, it allocates the data and sets the pointer accordingly I guess ?
https://github.com/KVM-VMI/kvm/blob/kvmi/tools/kvm/kvmi/include/kvmi/libkvmi.h#L106

Reply:
You're right. `kvmi-test.c` has memory leaks. You should call `free()`.